### DISCUSSION TITLE
How to use DEB packcage correctly?

### DISCUSSION THREAD

**Discussion by gordontcp** [fwupd/fwupd#5315] [Q&A] [answered]
I follow the instructions below,
`https://github.com/fwupd/fwupd/tree/main/contrib`
Execute the following command on Fedora x64 system:
```
OS=ubuntu-x86_64 ./generate_docker.py build
docker run --privileged -t -v `pwd`:/github/workspace fwupd-ubuntu-x86_64
```
The environment of this system is as follows:
`Linux fedora 6.0.11-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Dec 2 20:38:11 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux`
At this time, using docker, I can see the image as follows:
[x@fedora fwupd]$ sudo docker images
```
REPOSITORY       TAG             IMAGE ID       CREATED       SIZE
fwupd-ubuntu-x86_64 latest 4fd0d5ad18ea 2 days ago 144MB
```
Then pull the image on another ubuntu x64 system, I can see
```
x@xxx:~$ sudo docker images
REPOSITORY               TAG      IMAGE ID      CREATED SIZE
fwupd-ubuntu-x86_64 latest 4fd0d5ad18ea 2 days ago 144MB
```
The ubuntu system information is
```
x@xxx:~$ uname -a
Linux xxx 5.4.0-135-generic #152~18.04.2-Ubuntu SMP Tue Nov 29 08:23:49 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Finally I execute the docker run command and get the following result:
```
x@xxx:~$ sudo docker run fwupd-ubuntu-aarch64
docker: Error response from daemon: failed to create shim: OCI runtime create failed: runc create failed: unable to start container process: exec: "./contrib/ci/ubuntu.sh": stat ./contrib/ci/ubuntu. sh: no such file or directory: unknown.
ERRO[0001] error waiting for container: context canceled
```
My question is as follows:
   How to use the built deb package correctly?

Any suggestion is highly appreciated.
Thanks!

**Comment by hughsie** [comment]
@superm1 any ideas?

**Comment by superm1** [comment]
>  sudo docker run fwupd-ubuntu-aarch64

aarch64?  Are you trying to cross compile?

**Comment by gordontcp** [comment]
>sudo docker run fwupd-ubuntu-aarch64

Sorry for my wrong typing, the command should be:
```
sudo docker run fwupd-ubuntu-x86_64
docker: Error response from daemon: failed to create shim: OCI runtime create failed: runc create failed: unable to start container process: exec: "./contrib/ci/ubuntu.sh": stat ./contrib/ci/ubuntu.sh: no such file or directory: unknown.
ERRO[0002] error waiting for container: context canceled
```

`sudo docker run fwupd-ubuntu-aarch64` is another issue I would like to try! 
Thanks for your response!

**Comment by superm1** [comment, marked as answer]
I think the part missing is that the checkout needs to be bind-mounted into the container to be properly executed from the right place.
If you look at what we do in CI, we do this:

```
          docker run --privileged -e CI=true -t -v $GITHUB_WORKSPACE:/github/workspace docker.pkg.github.com/fwupd/fwupd/fwupd-${{matrix.os}}:latest
```

That is, bind-mount your checkout to /github/workspace in the container.

**Reply by superm1** [reply]
Our other CI job (circle) does it this way too:

```
docker run --privileged -e CI=true -t -v `pwd`:/github/workspace fwupd-ubuntu-x86_64
```

**Comment by gordontcp** [comment]
Thanks for your response!
I have some confusion. As far as I know, there are three steps to operate DEB package:
Step 1. prepare container
`OS=ubuntu-x86_64 ./generate_docker.py build`
Step 2. Build deb
`sudo docker run --privileged -e CI=true -t -v $GITHUB_WORKSPACE:/github/workspace  fwupd-ubuntu-x86_64`
Step 3. Launch container
`sudo docker run --privileged -e CI=true -i -t -v $GITHUB_WORKSPACE:/github/workspace -v /dev/:/dev/  fwupd-ubuntu-x86_64 bash`
where $GITHUB_WORKSPACE is /home/x/work/fwupd
Is there any mistake in the operation of steps 2 and 3 above?
（Note, I performed the action of step test 3 launch container in another x64 ubuntu 22.04 environment, but I did not see the fwupd directory being packaged.）

**Reply by superm1** [reply]
You don't actually need to launch the container (your step 3) unless you're debugging something.  Typically it's supposed to package binaries and spit them back out by your step 2.  If you're still not working properly, I suspect your bind mount should be to `/home/x/work` not `/home/x/work/fwupd`.

**Comment by gordontcp** [comment]
Thanks for your information.
I still have another question:
I have used the following docker hub package
`https://hub.docker.com/r/superm1/fwupd-issue-test`
I followed the instructions below:
`sudo docker run --privileged -i -t -v pwd:/build -v /dev/:/dev/superm1/fwupd-issue-test bash`
In the docker environment, I can see the fwupdtool file in `/usr/local/libexec/fwupd/` directory.
May I ask how this is done?
Are different parameters executed in my step2?
(Note , In my self-generated deb package, I don't see fwupdtool packaged into /usr/local/libexec/fwupd folder)
Thanks!

**Reply by superm1** [reply]
That thing is ancient and not at all updated. I suggest not using it.

**Comment by minzak** [comment]
```
me@elitebook /usr/local/src/fingerprint/fwupd-1.9.1/contrib $ OS=debian-x86_64 ./generate_docker.py build
bash: ./generate_docker.py: No such file or directory
```

Need another command:
```
me@elitebook /usr/local/src/fingerprint/fwupd-1.9.1/contrib $ OS=debian-x86_64 ./ci/generate_docker.py build
[+] Building 1.4s (10/10) FINISHED                                                                                                                                                                                                          
 => [internal] load build definition from Dockerfile                                                                                                                                                                                   0.0s
 => => transferring dockerfile: 1.63kB                                                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                      0.1s
 => => transferring context: 2B                                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/library/debian:testing                                                                                                                                                                      1.2s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                                                                                                                          0.0s
 => [1/5] FROM docker.io/library/debian:testing@sha256:116b23e7e027c7d6b011df8bf6c4673820706ae9f5c69ce966303da47f5d7cd6                                                                                                                0.0s
 => CACHED [2/5] RUN echo fubar > /etc/machine-id                                                                                                                                                                                      0.0s
 => CACHED [3/5] RUN apt update -qq &&  DEBIAN_FRONTEND=noninteractive apt install -yq --no-install-recommends bash-completion  fwupd-unsigned-dev  ca-certificates  clang-format  libcairo-dev  libcairo-gobject2  libjson-glib-dev   0.0s
 => CACHED [4/5] RUN apt install -yq --no-install-recommends python3-apt                                                                                                                                                               0.0s
 => CACHED [5/5] WORKDIR /github/workspace                                                                                                                                                                                             0.0s
 => exporting to image                                                                                                                                                                                                                 0.0s
 => => exporting layers                                                                                                                                                                                                                0.0s
 => => writing image sha256:479a97e38aa930aa11f85a92fe1af771eb996e1c7e83183c283d2892d06ae5be                                                                                                                                           0.0s
 => => naming to docker.io/library/fwupd-debian-x86_64      
```

And next after run command like:
```
sudo docker run --privileged -e CI=true -t -v $(pwd):/github/workspace  fwupd-debian-x86_64
```
No where placed any DEB files! 
But here is present directory that have almost all for build DEB files, see similar instructions, but here needs a lot modifications.
https://community.frame.work/t/responded-fingerprint-scanner-compatibility-with-linux-ubuntu-fedora-etc/1501/131

Any worked solution to get real DEB files?

**Reply by superm1** [reply]
Are you running from the root of the checkout?  That's where you should be launching everything, including building the container.
